src/clj/disaster/middleware/oauth2_jwt.clj

```diff
@@ -0,0 +1,197 @@
+(s/def ::jwk (s/keys :req-un [::kty ::e ::n ::kid]))
+(s/def ::public-key keys/public-key?)
+(s/def ::private-key keys/private-key?)
+(s/def ::key (s/keys :req-un [::public-key]
+                     :opt-un [::private-key]))
+(s/def ::key-store (s/map-of ::kid
+                             ::key))
+(s/def ::resource (s/with-gen #(or (instance? java.net.URL %)
+                                   (instance? (class (char-array "")) %))
+                    ;; Always use local resources to avoid spamming actual servers
+                    #(s/gen #{(resource "jwks.json")
+                              (resource "jwks-other.json")})))
+(s/def ::jwks-url (s/with-gen (s/or :url  :invetica.uri/absolute-uri-str
+                                    :resource ::resource)
+                    ;; Always use local resources to avoid spamming actual servers
+                    #(s/gen #{(resource "jwks.json")
+                              (resource "jwks-other.json")})))
+
+(s/fdef jwks-edn->keys
+  :args (s/cat :jwks (s/coll-of ::jwk :type vector?))
+  :ret ::key-store)
+
+(defn- jwks-edn->keys
+  "Transform a vector of json web keys into a map of kid -> key pairs where each key is a map
+  of :public-key and optionally :private-keys."
+  [json-web-keys]
+  (->> json-web-keys
+       :keys
+       (filter #(= (:kty %) "RSA"))
+       (group-by :kid)
+       (fmap first)
+       (fmap #(assoc {}
+                     :public-key (buddy-jwk/jwk->public-key %)
+                     :private-key (buddy-jwk/jwk->private-key %)))))
+
+(s/fdef fetch-keys
+  :args (s/cat :jwks-url ::jwks-url)
+  :ret  (s/with-gen ::key-store
+          #(s/gen #{(->> (resource "jwks.json")
+                         slurp
+                         ((fn [jwks-string] (json/read-str jwks-string :key-fn keyword)))
+                         jwks-edn->keys)})))
+
+(defn- fetch-keys
+  "Fetches the jwks from the supplied jwks-url and converts to java Keys.
+  Returns a map keyed on key-id where each value is a RSAPublicKey object"
+  [jwks-url]
+  (log/debug "Fetching keys from jwks-url" jwks-url)
+  (try  (->> jwks-url
+             slurp
+             (#(json/read-str % :key-fn keyword))
+             jwks-edn->keys)
+        (catch Exception e (do (log/error "Could not fetch jwks keys")
+                               false))))
+
+
+;; Atom to hold the public and private keys used for signature validation in memory for
+;; caching purposes. The atom holds a clojure map with kid -> key pairs. Each key is a
+;; clojure map containing a :public-key and optionally a :private-key.
+(defonce keystore
+  (atom {}))
+
+
+(defn- resolve-key
+  "Returns java.security.Key given key-fn, jwks-url and :key-type in jwt-header.
+  If no key is found refreshes"
+  [key-type jwks-url jwt-header]
+  (log/debug "Resolving key" jwt-header "from jwk cache for" jwks-url)
+  (let [key-fn (fn [] (get-in @keystore [jwks-url (:kid jwt-header) key-type]))]
+    (if-let [key (key-fn)]
```

Comment:
you can just concat both the definition and function call for `key-fn` in one go. Optionally, prob move this outside and supply it as an argument if you want to change how to get the key later. For compatibility, just have another overload function that use the `key-fn` as the default argument. So something like:

```
(defn- resolve-key
    [key-type jwks-utl jwt-header key-fn]
    (...)
    [key-type jwks-utl jwt-header]
    (recur key-type jwks-util jwt-header key-fn))
```